### PR TITLE
Support DocService for AbstractHttpService

### LINKED ISSUE #1957: Provide a way to generate doc service using Route API
Currently, `DocService` only supports to generate specification from annotated service. Even though the restriction is described in the documentation, users get confused when their REST API written by [Route API]((https://github.com/line/armeria/blob/07f3571b55b7f19f1e66596cb303543b6ce45fab/core/src/main/java/com/linecorp/armeria/server/Route.java#L32) doesn't appear on the doc service. It is also useful when Armeria supports Kotlin DSL API.

### PR DESCRIPTION
**Pull request by heowc**
Motivation:

This is to expose `AbstactHttpService` to `DocService` by adding `HttpDocServicePlugin`.

Modifications:

- Change the logic to find each service based on the `Route`s.
- Give priority to each plugin
- Add `HttpDocServicePlugin`
- Add `AbstractDocServicePlugin` for base
- Add test

Result:

- Closes #1957

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/internal/server/HttpDocServicePlugin.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+    @VisibleForTesting
+    static ServiceSpecification generate(Map<Class<?>, String> serviceDescription,
+                                         Map<Class<?>, Set<MethodInfo>> methodInfos) {
+        final Set<ServiceInfo> serviceInfos = methodInfos
+                .entrySet().stream()
+                .map(entry -> {
+                    final Class<?> service = entry.getKey();
+                    return new ServiceInfo(service.getName(), entry.getValue(),
+                                           serviceDescription.get(service));
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/server/HttpDocServicePlugin.java:132]: Isn't `serviceDescription.get(service)` always null?

File: core/src/main/java/com/linecorp/armeria/internal/server/HttpDocServicePlugin.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        return HttpDocServicePlugin.class.getSimpleName();
+    }
+
+    private static void addMethodInfo(Map<Class<?>, Set<MethodInfo>> methodInfos, String hostnamePattern,
+                                      AbstractHttpService service, Route route, HttpMethod httpMethod) {
+        final EndpointInfo endpoint = endpointInfo(route, hostnamePattern);
+        final Class<?> clazz = service.getClass();
+        final String methodName = METHOD_NAMES.get(httpMethod);
+        final MethodInfo methodInfo = new MethodInfo(
+                methodName, HTTP_RESPONSE, ImmutableList.of(), ImmutableList.of(), // Ignore exceptions.
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/server/HttpDocServicePlugin.java:118]: How about defining a static final field list that has `ServiceRequestContext` and `HttpRequest` and passes it here as a parameter?

File: core/src/main/java/com/linecorp/armeria/internal/server/docs/NoopDocServicePlugin.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import com.linecorp.armeria.server.Service;
+import com.linecorp.armeria.server.ServiceConfig;
+import com.linecorp.armeria.server.docs.DocServiceFilter;
+import com.linecorp.armeria.server.docs.DocServicePlugin;
+import com.linecorp.armeria.server.docs.ServiceSpecification;
+
+/**
+ * A {@link DocServicePlugin} implementation that generate empty {@link ServiceSpecification}.
+ */
+public final class NoopDocServicePlugin implements DocServicePlugin {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/internal/server/docs/NoopDocServicePlugin.java:33]: Let's use enum singleton. 😄 
You might refer this class: https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/internal/common/NoopKeepAliveHandler.java

**Review by minwoox — COMMENTED** [review]
Thanks, @heowc. Left some minor comments. 😄 

> Closes Provide a way to generate doc service using Route API #1957

By the way, this does not close #1957 because it does not handle services that are created using `Route` API. 😄

**Comment by ikhoon** [comment]
> > Closes Provide a way to generate doc service using Route API #1957
> 
> By the way, this does not close #1957 because it does not handle services that are created using `Route` API. 😄

Commented random thoughts in https://github.com/line/armeria/pull/3661#issuecomment-871033094

**Comment by heowc** [comment]
> Thanks, @heowc. Left some minor comments. 😄
> 
> > Closes Provide a way to generate doc service using Route API #1957
> 
> By the way, this does not close #1957 because it does not handle services that are created using `Route` API. 😄

Hmm... Then wouldn't this PR be useful? If so, I'll close this PR and think again. 🤔

**Comment by minwoox** [comment]
> Hmm... Then wouldn't this PR be useful? If so, I'll close this PR and think again. 🤔

I thought we could provide another doc service plugin for the route.
However, if we implement the one, then we might not need this `HttpDocSerivcePlugin`. 🤔 
How about just leaving this PR as it is and continuing to discuss it on the comment https://github.com/line/armeria/pull/3661#issuecomment-871033094?

**Comment by heowc** [comment]
> How about just leaving this PR as it is and continuing to discuss it on the comment #3661 (comment)?
That's a good idea. 😎

**Comment by heowc** [comment]
The intended direction is different, so I will close the PR. 😆

**Comment by minwoox** [comment]
> The intended direction is different, so I will close the PR. 😆

Yeah, let's close this PR. By the way, thanks a lot for your constant effort to improve our DocService! 😄

**Pull request closed without merging**